### PR TITLE
[Table] Make row and column heights/widths sparse arrays types correct

### LINKED ISSUE #879: Table's `columnWidths` typings disagree with the doc comment
[From `Table`](https://github.com/palantir/blueprint/blob/master/packages/table/src/table.tsx#L92..L98):

    /**
     * A sparse number array with a length equal to the number of columns. Any
     * non-null value will be used to set the width of the column at the same
     * index. Note that if you want to update these values when the user
     * drag-resizes a column, you may define a callback for `onColumnWidthChanged`.
     */
    columnWidths?: number[];

If the doc comment is right, the type should be `(number | null | undefined)[]` (assuming by "non-null" is means "non-null, non-undefined"). Otherwise, the doc comment should be changed to agree with the typings.

### PR DESCRIPTION
**Pull request by gscshoyru**
#### Changes proposed in this pull request:

In a world with strict-null-checks, the typing is wrong -- make it so the typing actually supports sparse arrays in a strict null check world.

### CODE REVIEW

File: packages/table/src/table.tsx [hunk 1]
```diff
@@ -95,7 +95,7 @@ export interface ITableProps extends IProps, IRowHeights, IColumnWidths {
      * index. Note that if you want to update these values when the user
      * drag-resizes a column, you may define a callback for `onColumnWidthChanged`.
      */
-    columnWidths?: number[];
+    columnWidths?: Array<number | null | undefined>;
```
Comment (adidahiya) [inline packages/table/src/table.tsx]: does we need to allow null? how about we just allow `null | undefined`?

Comment (gscshoyru) [inline packages/table/src/table.tsx]: That's probably fine.

Comment (gscshoyru) [inline packages/table/src/table.tsx]: Well, hm. It says non-null value in the docs. Should I change that to match as well? More to the point, any reason _not_ to allow null?

Comment (adidahiya) [inline packages/table/src/table.tsx]: To discourage use of `null` in general. If this is bothersome to consumers we can change it to be more lenient later.

Comment (themadcreator) [inline packages/table/src/table.tsx]: Well if we intend to enable strict null checks then won't we no longer be able to write

```javascript
if (columnWidth[123] == null) {
   // ...
}
```
If that's true then it'd be nicer to only have to do `=== undefined` versus checks for both

Comment (themadcreator) [inline packages/table/src/table.tsx]: But whichever way you go, this approach looks good

Comment (adidahiya) [inline packages/table/src/table.tsx]: nothing stops you from writing `== null` even if this type changes to `Array<number | undefined>`. what _would_ stop you from writing that is the [no-null-keyword](https://palantir.github.io/tslint/rules/no-null-keyword/) rule. see [here](https://pl.ntr/null-vs-undefined) for my thoughts on null vs undefined

**Comment by adidahiya** [comment]
ok I guess I should've requested changes and blocked this merge... please see my suggestion above and justify why you want to include `null` if you disagree

File: packages/table/src/table.tsx (hunk 1 above)
Comment (seansfkelley) [inline packages/table/src/table.tsx]: FWIW, as the filer of the issue in question, I think accepting `null` is the right thing to do (the Robustness Principle and all that). In my experience, opinions are split in the community on whether `null`, `undefined` or both are acceptable as the "null type", and purposely not accepting one of them as a library author is poor devex. If people want to lint out their usages of `null`, that's their prerogative.

**Pull request merged**